contracts/contracts/strategies/GeneralizedUniswapV3Strategy.sol

```diff
@@ -0,0 +1,1144 @@
+        if (token1Balance > 0) {
+            token1Contract.safeTransfer(vaultAddress, token1Balance);
+            emit Withdrawal(token1, token1, token1Balance);
+        }
+    }
+
+    function _getToken1ForToken0(uint256 amount0) internal {}
+
+    /**
+     * @dev Checks if there's enough balance left in the contract to provide liquidity.
+     *      If not, tries to pull it from reserve strategies
+     * @param desiredAmount0 Minimum amount of token0 needed
+     * @param desiredAmount1 Minimum amount of token1 needed
+     */
+    function _ensureAssetBalances(
+        uint256 desiredAmount0,
+        uint256 desiredAmount1
+    ) internal {
+        IVault vault = IVault(vaultAddress);
+
+        // Withdraw enough funds from Reserve strategies
+        uint256 token0Balance = IERC20(token0).balanceOf(address(this));
+        if (token0Balance < desiredAmount0) {
+            vault.withdrawAssetForUniswapV3(
+                token0,
+                desiredAmount0 - token0Balance
+            );
+        }
+
+        uint256 token1Balance = IERC20(token1).balanceOf(address(this));
+        if (token1Balance < desiredAmount1) {
+            vault.withdrawAssetForUniswapV3(
+                token1,
+                desiredAmount1 - token1Balance
+            );
+        }
+
+        // TODO: Check value of assets moved here
+    }
+
+    function _ensureAssetsBySwapping(
+        uint256 desiredAmount0,
+        uint256 desiredAmount1,
+        uint256 swapAmountIn,
+        uint256 swapMinAmountOut,
+        uint160 sqrtPriceLimitX96,
+        bool swapZeroForOne
+    ) internal {
+        require(!swapsPaused, "Swaps are paused");
+        IERC20 t0Contract = IERC20(token0);
+        IERC20 t1Contract = IERC20(token1);
+
+        uint256 token0Balance = t0Contract.balanceOf(address(this));
+        uint256 token1Balance = t1Contract.balanceOf(address(this));
+
+        uint256 token0Needed = desiredAmount0 > token0Balance
+            ? desiredAmount0 - token0Balance
+            : 0;
+        uint256 token1Needed = desiredAmount1 > token1Balance
+            ? desiredAmount1 - token1Balance
+            : 0;
+
+        if (swapZeroForOne) {
+            // Amount available in reserve strategies
+            uint256 t1ReserveBal = IStrategy(poolTokens[token1].reserveStrategy)
+                .checkBalance(token1);
+
+            // Only swap when asset isn't available in reserve as well
+            require(
+                token1Needed > 0 && token1Needed < t1ReserveBal,
+                "Cannot swap when the asset is available in reserve"
+            );
+            // Additional amount of token0 required for swapping
+            token0Needed += swapAmountIn;
+            // Subtract token1 that we will get from swapping
+            token1Needed -= swapMinAmountOut;
+
+            // Approve for swaps
+            t0Contract.safeApprove(address(swapRouter), swapAmountIn);
```

Comment:
should we do this only once (maybe on contract initialisation) rather than on every swap

Reply:
Yep. I added that as a comment in the new contract that I pushed last night